cpp/src/components/connectivity.cu

```diff
@@ -41,190 +40,71 @@ namespace detail {
-  CUGRAPH_EXPECTS(col_indices_(graph) != nullptr, "Invalid API parameter: col_indices_(graph) is empty");
-  
-  CUGRAPH_EXPECTS(labels->data != nullptr, "Invalid API parameter: labels data is NULL");
-  CUGRAPH_EXPECTS(verts->data != nullptr, "Invalid API parameter: verts data is NULL");
-  
-  auto type_id = graph->adjList->offsets->dtype;
-  CUGRAPH_EXPECTS( type_id == GDF_INT32 || type_id == GDF_INT64, "Unsupported data type: graph must be int32 or int64");
-  CUGRAPH_EXPECTS( type_id == graph->adjList->indices->dtype,  
-    "Unsupported data type: offsets and indices in the adjacency list have different types");
+  using ByteT = unsigned char;//minimum addressable unit
   
-  //TODO: relax this requirement:
-  //
-  CUGRAPH_EXPECTS( type_id == labels->dtype, "Unsupported data type: labels->dtype must be int32 or int64");
-
-  IndexT* p_d_labels = static_cast<IndexT*>(labels->data);
-  IndexT* p_d_verts = static_cast<IndexT*>(verts->data);
+  CUGRAPH_EXPECTS(graph.offsets != nullptr, "Invalid API parameter: graph.offsets is nullptr");
+  CUGRAPH_EXPECTS(graph.indices != nullptr, "Invalid API parameter: graph.indices is nullptr");
   
-  const IndexT* p_d_row_offsets = row_offsets_(graph);
-  const IndexT* p_d_col_ind = col_indices_(graph);
-
-  IndexT nnz = nnz_(graph);
-  IndexT nrows = nrows_(graph);//static_cast<IndexT>(graph->adjList->offsets->size) - 1;
+  VT nrows = graph.number_of_vertices;
   
-  if( connectivity_type == CUGRAPH_WEAK )
-    {
-      // using VectorT = thrust::device_vector<IndexT>;
-      // VectorT d_ro(p_d_row_offsets, p_d_row_offsets + nrows + 1);
-      // VectorT d_ci(p_d_col_ind, p_d_col_ind + nnz);
-
-#ifdef _DEBUG_WEAK_CC
-      IndexT last_elem{0};
-      cudaMemcpy((void*)(&last_elem), p_d_row_offsets+nrows, sizeof(IndexT), cudaMemcpyDeviceToHost);
-      std::cout<<"############## "
-               <<"nrows = "<<nrows
-               <<"; nnz = "<<nnz
-               <<"; nnz_ro = "<<last_elem
-               <<"; p_d_labels valid: "<<(p_d_labels != nullptr)
-               <<"; p_d_row_offsets valid: "<<(p_d_row_offsets != nullptr)
-               <<"; p_d_col_ind valid: " << (p_d_col_ind != nullptr)
-               <<"\n";
-      
-      std::cout<<"############## d_ro:\n";
-      print_v(d_ro, std::cout);
-
-      std::cout<<"############## d_ci:\n";
-      print_v(d_ci, std::cout);
-#endif
-
+  if (connectivity_type == cugraph_cc_t::CUGRAPH_WEAK) {
       //check if graph is undirected; return w/ error, if not?
       //Yes, for now; in the future we may remove this constraint; 
       //
-      bool is_symmetric = cugraph::detail::check_symmetry(nrows, p_d_row_offsets, nnz, p_d_col_ind);
-#ifdef _DEBUG_WEAK_CC
-      std::cout<<"############## "
-               <<"; adj. matrix symmetric? " << is_symmetric
-               <<"\n";
-#endif
+      bool is_symmetric = cugraph::detail::check_symmetry(graph.number_of_vertices, graph.offsets, graph.number_of_edges, graph.indices);
```

Comment:
Opinions on this?

The python bindings symmetrize the graph if it is not already symmetrized.  I haven't work through all of the paths in the python code, but it seems a user would have to work hard to get here without a symmetrized graph.

The C++ path would clearly allow that.

I'm inclined not to pay the cost of this check.

Reply:
yes, I think this is a good topic for the graph alignment meeting, but for expensive error checking, w need to have a policy.

If we cannot finish due to any sort of errors, it is better to throw exception as early as possible. So, I assume it is better to perform expensive error checking on python side right after user calls a cugraph function (with an option to disable checking on users' own risk if they are confident).

And for the C++ side, I think it is a good practice to provide a flag to enable expensive checking for debugging purpose, but for the default case, better not repeat expensive error checking.

Reply:
This check was already in the code, I left it in hoping to have this conversation.

The "danger" of doing a WCC against and unsymmetrized graph is that you'll get a meaningless answer.  I'd have to think through this a bit, but it seems like with this implementation the "danger" is that it might take more iterations of the loop to compute the correct answer.  Which is a silly reason to fail

Reply:
OK, as long as we provide an option to check the input, if the caller decides to skip it for performance reason, the caller is doing it in the caller's own risk, so I am less worried about that case.

I think if the caller asks to skip input check and the input violates pre-conditions, we can say this function's behavior is "undefined" at least in the C++ level. (So, if this is un-acceptable, need to check this in python level or do not disable input check).

Reply:
I'm inclined to remove this entirely.

1. The python bindings explicitly symmetrize the graph if called with a DiGraph, so the python caller should be passing a graph that has already been symmetrized.
1. A C++ caller could create a graph that is not symmetric.  If that happens, the weak_cc implementation may take extra iterations to generate a solution.  But I believe it will actually generate the correct solution... just requiring extra iterations.